### PR TITLE
Share statecharts between models

### PR DESCRIPTION
**Pull request by lmignon**

### CODE REVIEW

File: statechart/models/ir_model.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+class IrModel(models.Model):
+
+    _inherit = 'ir.model'
+
+    statechart_id = fields.Many2one(
+        'statechart',
+    )
+
+    @api.constrains('statechart_id')
+    def _chec_statechart_id(self):
```
Comment (sbidoul) [inline statechart/models/ir_model.py]: _chec*k*_statechart_id

**Comment by sbidoul** [comment]
A typo otherwise looks good.

**Comment by lmignon** [comment]
@sbidoul The last commit should interest you.

**Comment by lmignon** [comment]
@sbidoul The last commit requires a deep review...

File: statechart_demo_purchase/demo/statechart.xml [hunk 2]
```diff
@@ -79,4 +80,35 @@ statechart:
...
+    states:
+      - name: draft
+        transitions:
+          - event: button_grand_parent
+        </field>
+    </record>
+
+    <record model="statechart" id="statechart_child2_demo">
+        <field name="name">test.inherit.child2</field>
+        <field name="model_ids" eval="[(4, ref('statechart_demo_purchase.model_test_inherit_child2'))]"/>
```
Comment (sbidoul) [inline statechart_demo_purchase/demo/statechart.xml:101]: What should happen when a subclass redefines the statechart of a parent class?

I suppose it should replace it completely. The developer then being responsible to ensure the statechart of the subclass is compatible with the statechart of the parent.

File: statechart_demo_purchase/tests/test_inherit.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            self.child2.button_grand_parent()
+        with self.assertRaises(AttributeError):
+            self.child2.sc_button_grand_parent_allowed
+        # check that the statechart is inherited for child1
+        self.assertEqual(
+            self.child1.sc_interpreter.statechart.name,
+            'test.inherit.grand.parent'
+        )
+        self.assertTrue(self.child1.sc_button_grand_parent_allowed)
+        self.child1.button_grand_parent()
```
Comment (sbidoul) [inline statechart_demo_purchase/tests/test_inherit.py:40]: fails here, but it's going to be hard to have L40 and L30 pass at the same time

File: statechart_demo_purchase/demo/statechart.xml (hunk 2 above)
Comment (lmignon) [reply statechart_demo_purchase/demo/statechart.xml:101]: I also suppose that it should replace it completely.

**Comment by sbidoul** [comment]
@lmignon can you have a look at #6 which continues on top of this one?

**Comment by lmignon** [comment]
superseded by #8

**Pull request closed without merging**